PeggyJV/cellar-contracts
# Design, Develop, and Test Convex Adaptors

## Core changes:

This PR encompasses the development and unit testing of the Convex Adaptor that allows:

1. Depositing Curve LPTs
2. Withdrawing Curve LPTs
3. Claiming Rewards

There are several different phases / scopes. They are as follows:
- Scope 1 (for Curve-Convex Market): `SupplyLPT` Functionality
- Scope 2 (for Curve-Convex Market): `WithdrawLPT` Functionality
- Scope 3 (for Curve-Convex Market): `ClaimRewards` Functionality
- Scope 4: Creation of a Frax-Convex Market Adaptor with similar functionality as that done above with Curve-Convex Markets

Extra: HealthFactor Logic smart contracts that likely will be used by the `CollateralMorphoBlueAdaptor.sol` and `DebtMorphoBlueAdaptor.sol`

---
<details markdown='1'><summary> ℹ️ Context </summary>

### Context:
tldr of general resaerch; General context on rewards setup of Convex to help scope out work - A user/contract can collect rewards by:
- depositing lpt & staking cvxLPT (not sure if there is much benefit w/ just holding cvxLPT)
    - boosted CRV, liquidity mining rewards
- staking cvxCRV (aka staking CRV effectively for the protocol)
    - 3crv gov fee, tx fees, share of boosted CRV from LPs, liquidity mining rewards (as CVX)
- staking CVX
    - performance fees (assuming as CVX)
- depositing CRV and getting cvxCRV
    - not sure if we are going to get any rewards for just holding cvxCRV

Functionality listed in priority:

1. depositing lpt & staking cvxLPT
2. withdrawing & unwrapping cvxLPT to CRVLPT
3. claiming rewards from all reward contracts (TODO: confirm that if the cellar is staking CRV(cvxCRV), CVX, then if it calls withdraw and unwrap, and sets the bool to claim rewards that it would get all rewards associated to the cellar)
4. staking CVX
5. staking CRV

</details>

---

<details markdown='1'><summary> ℹ️ TODOs </summary>

### General TODO:
- [x] Research Curve-Convex Markets and codebase
- [x] Review scope with Crispy before writing up code
- [ ] Research FRAX-Convex Markets and codebase

### Scope 1: `SupplyLPT` Functionality
- [x] Stub out code
- [x] Add strategist functions interacting w/ external protocol
- [x] Add base functions
- [ ] Reformat and remove unneeded comments and TODOs
- [ ] Full PR Review from Crispy. 
- [ ] PR Review and refine
- [ ] Finalize

### Scope 2: `WithdrawLPT` Functionality
- [x] Stub out code
- [x] Add strategist functions interacting w/ external protocol
- [x] Add base functions
- [ ] Reformat and remove unneeded comments and TODOs
- [ ] Full PR Review from Crispy. 
- [ ] PR Review and refine
- [ ] Finalize

### Scope 3: `ClaimRewards` Functionality
- [x] Stub out code
- [x] Add strategist functions interacting w/ external protocol
- [x] Add base functions
- [ ] Reformat and remove unneeded comments and TODOs
- [ ] Full PR Review from Crispy. 
- [ ] PR Review and refine
- [ ] Finalize

</details>

## Notes:
- Research: Will review Convex current codebase, app, and docs before looking at the previous PRs.
- Reviews will be done for the pseudo-code for above functionalities and then for full implementation.

---
<details markdown='1'><summary> ℹ️ Live Discussion Topics (discussed during the Draft PR stage) </summary>

- Boosting Curve positions on our own, that is up to the curve adaptor design. Likely not as that adds much more complexity dealing with ve voting.
- Boosting / gaining more APR via interacting with veCVX, that is also out of scope for now.
- Exposure to any past risks w/ Curve / Convex: TBD
- Claiming Extra Rewards outside of the BaseReward contracts:
    - `baseRewardPool.getReward( address, bool )` should cover claiming a user's accrued rewards (inclusive of extra rewards). There is a `stash` contract for each baseRewardPool contract that holds onto extra rewards external to Convex. The `bool` within `getReward()` allows one to claim the reward tokens from the stash contract.
    - TODO: confirm this with tests.

</details>
---
<details markdown='1'><summary> ℹ️ Questions & References </summary>

Technical Questions:
- Is there a way to query the Booster.sol for the associated BaseRewardsPool.sol? If not... then I guess we use the subgraph?
- Is the Booster contract the main gateway to deposit and withdraw LPTs, and stake them, and then the BaseRewardsPool used to withdraw and stake rewards tokens? So it looks like BaseRewardsPool uses `IDeposit` from SNX, and from checking around it connects to the `Booster` contract and withdraws the LPTs.
    - So for withdrawing, it could make sense to use `withdrawAndUnwrap`

---
## References:
1. Convex Docs: https://docs.convexfinance.com/convexfinanceintegration/
2. Convex Codebase: https://github.com/convex-eth // TODO: not sure why it's not working
3. Prior PR from zkoranges here: TODO:
4. Prior branch from Crispy: TODO:

</details>

---

<details markdown='1'><summary> ℹ️ Extra Notes & Tests to Carry Out </summary>

1. Tests to Carry Out:
- Check if cellar gets receipt tokens from depositing into convex any LPTs.
    - If yes, then write tests confirming we get correct amount. Maybe TODO: of implementing the check onchain in the same tx.
                // TODO: not sure if we actually get a cvx recipt token... I think it might actually just be internal. One way to check... is to go to the actual deployed convex booster contract and see if ppl receive anything. And to test it.

---

General Research Notes/Thoughts:

We want functionality that covers:

Within `Booster.sol` (the main deposit contract for LP tokens)
- `deposit()` LPTs & receive convex receipt token (deposit into a pool "n" and receive the deposit token)
   - maybe `depositAll()` LPTs & receive convex receipt tokens (this function deposits all lp tokens for pool "n" and stake into the rewards contract)
- `withdraw()` LPTs & burn convex receipt token
    - maybe `withdrawAll()` which withdraws all from pool "n"
- `stake` convex receipt tokens into Convex for more rewards

Within `BaseRewardPool.sol` - https://docs.convexfinance.com/convexfinanceintegration/baserewardpool
- `baseRewardPool.getReward( address, bool )` to claim baseRewardPool rewards, and extra rewards (signified by passed param of `bool` - true indicates to grab any extra rewards from `stash` contract)
- `stake()` convex receipt tokens (in case they were not staked for some reason when initially depositing LPTs).
    - Use baseRewardPool.stake( amount ) to stake a given amount for your address
    - Use baseRewardPool.stakeAll() to stake all tokens for your address
    - Use baseRewardPool.stakeFor( address, amount ) to stake on behalf of another address (*This transfers ownership of the tokens!)
- `withdraw()`
    - Use baseRewardPool.withdraw( amount, bool ) to withdraw a specific amount. bool is to also claim while withdrawing
    - Use baseRewardPool.withdrawAll( bool ) to withdraw all tokens staked. bool is to also claim while withdrawing
    - Use baseRewardPool.withdrawAndUnwrap( amount, bool ) to withdraw a specific amount of tokens AND also unwrap the tokens back into the underlying LP tokens. bool is to also claim while withdrawing
    - Use baseRewardPool.withdrawAllAndUnwrap(bool claim) to withdraw all tokens AND also unwrap the tokens back into the underlying LP tokens. bool is to also claim while withdrawing

Functionality I am unsure if it's in the scope of this PR:
- Staking CRV
- Staking cvxCRV (liquid veCRV)


</details>

<!--- example - Design decisions made in PR>
<!--- example - Remember running `forge fmt` before committing the PR>
<!--- example - Reference for how to run relevant tests i.e. forge test --match-contract X --fork-url $RPC_URL -vvv>
